### ISSUE TITLE
Carrying out the LiteDRAM standalone core initialization manually, through wishbone ctrl interface

### ISSUE THREAD

**Issue by dinaabdelbaky** [enjoy-digital/litedram#327]
![image](https://user-images.githubusercontent.com/123503078/221815123-a428908e-46fa-4a38-9362-357b53c69699.png)
Hi,
I am new to the liteDRAM. I generated the core in a standalone mode, targeting Artix device, I also generated the corr. csv containing the csr addresses. I am using the core with Micron DDR3 model and a testbench to carryout come simulations and explorations of memory operations.
I checked out the generated sdram_phy.h it includes init_sequence, I tried to carry out the above sequence by writing to the corresponding cs registers. whoever I didn't see anything happening on the DFI interface or the memory interface.
Please point out what I am missing.
P.S: I have seen some similar questions, and some of the answers points out to writing data from an init file, but I don't really understand it, so please explain how to correctly initialize the core and the memory device.
Thanks in advance!

**Comment by dinaabdelbaky** [comment]
Hi,
I am new to the liteDRAM. I generated the core in a standalone mode, targeting Artix device, I also generated the corr. csv containing the csr addresses. I am using the core with Micron DDR3 model and a testbench to carryout come simulations and explorations of memory operations.
I checked out the generated sdram_phy.h it includes init_sequence, I tried to carry out the above sequence by writing to the corresponding cs registers. whoever I didn't see anything happening on the DFI interface or the memory interface.
Please point out what I am missing.
P.S: I have seen some similar questions, and some of the answers points out to writing data from an init file, but I don't really understand it, so please explain how to correctly initialize the core and the memory device.
Thanks in advance!

**Comment by yyx32** [comment]
Hi @dinaabdelbaky ，are you generated core with gen.py script?
Can you tell me how to get corr.csv and sdram_phy.h ?

**Comment by dinaabdelbaky** [comment]
Hi @yyx32 Yes, I generated the core following those steps:
The addresses of the registers would be in the csv file.
The sdram_phy.h is generated under "output_drisoftware\include\generated"
![image](https://github.com/enjoy-digital/litedram/assets/123503078/0685957f-0170-4e5e-b5d7-34b71e7db4ce)

**Comment by dinaabdelbaky** [comment]
And I have worked out the issue I have mentioned above, the problem was not the sequence, I didn't notice that the adresses in the CSV were 32 bit wide and the wishbone interface was less wider. A simple fix to that is to take the address of the csr register in the generated csv and do right bit shifts, the amount of bit shift is equivalent 32-wishbone width

**Comment by yyx32** [comment]
I will try it.Thanks bro.

**Comment by gvillanovanm** [comment]
> And I have worked out the issue I have mentioned above, the problem was not the sequence, I didn't notice that the addresses in the CSV were 32 bit wide and the wishbone interface was less wide. A simple fix to that is to take the address of the csr register in the generated csv and do right bit shifts, the amount of bit shift is equivalent 32-wishbone width.

Nice @dinaabdelbaky ! So, were you able to use the DDR3 manually? Because I'm trying the same thing.

**Comment by dinaabdelbaky** [comment]
@gvillanovanm Yes, I managed to get to work. I used the LiteDRAM core plus a DDR3 memory model available from micron as the DUT, and for the Testbench I used Cocotb. I hope this task is going smooth for you.
